### ISSUE TITLE
Usage of Proxy

### ISSUE THREAD

**Issue by mcollina** [feugy/mini-service#5]
What are the benefits of using Proxy?

Accessing something through proxy is 20x slower than doing the same through a standard object.

**Comment by feugy** [comment]
Following [this explanation](https://github.com/feugy/mini-service/issues/4#issuecomment-253815946)

TLDR; keep in mind the idea of supporting both monolih/µService deployements

Client code within A application will invoke B apis:

``` javascript
// in A
const client = require('B')()
client.init().then(() => {
  client.add(10, 5)
})
```

When B is configured as monolith, it's easy: the client object has one property per API exposed by B (`add()` in this example).
The `init()` function is responsible to create those properties.

Now B is deployed as a µService, and you configure A to use the remote B:

``` javascript
// in A
const client = require('B')({remote: 'http://b-service')
client.init().then(() => {
  client.add(10, 5)
})
```

B already expose [a kind of descriptor](https://github.com/feugy/mini-service/blob/4ccc6d867f1a2d3bb0c0d5bfbee290c9aa2da4b9/lib/server.js#L44) which indicates all the exposed API. 

This allow A's client to fetch the exposed API (with their respective signature), and creates function properties accordingly.

In the [earliest releases of mini-service](https://github.com/feugy/mini-service/commit/c694898bb211981df9ed8c5343dbbb6a9790cbb9), the `init()` method was responsible to do this. 
But this as a huge drawback: B remote service must be reachable when A's client init() is invoked, and in real life, this is rarely the case.

The usage of Proxy allows to defer the descriptor fetching and the client property creation, until A actually use one of B exposed API.

The 20x overhead exists, but only for the first call, because once the remote descriptor is fetched, properties are [created](https://github.com/feugy/mini-service/blob/master/lib/utils/register.js#L98) and [attached to the client (`target`) object](https://github.com/feugy/mini-service/blob/master/lib/client.js#L57). Further calls won't ever use the Proxy mechanism.

**Comment by mcollina** [comment]
It's really a nice use of proxy!!! Thanks for the explanation. I'm convinced. It would probably be nice to run 0x here to check everything is smooth http://npm.im/0x.

**Comment by mcollina** [comment]
Even if you reattach them on the target, you will incur in the 20x penalty. https://github.com/feugy/mini-service/blob/master/lib/client.js#L45-L48

**Comment by feugy** [comment]
Absolutely true, all my apologies.
This PR #6  should fix it.

**Comment by feugy** [comment]
PR #6 slightly improved the situation, and a benchmark allowed to measure the gap between the proxified mini-service client and its direct equivalent.

Proxy overhead is bellow 25%, which is to me an acceptable cost, considering that developers don't need to care about their client initialisation.
